### PR TITLE
Add forward AD for torch.linalg.eigh

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This PR adds forward mode differentiation for `torch.linalg.eigh` and a few other functions required for tests to pass.

For some reason running tests for `torch.linalg.eigvalsh` and complex `torch.linalg.eigh` hangs. These tests are skipped for now.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @jianyuh @mruberry @heitorschueroff @walterddr @IvanYashchuk @xwang233

### CODE REVIEW

**Comment by IvanYashchuk** [comment]
@albanD, CI fails with `NotImplementedError: Trying to use forward AD with _reshape_alias that does not support it.`. Do you have an idea how to enable forward AD for `_reshape_alias`? There's no such entry in derivatives.yaml file.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (Lezcano) [inline torch/csrc/autograd/FunctionsManual.cpp:2417]: It's better to divide by `F` than to compute the inverse explicitly an then multiply.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (Lezcano) [inline torch/csrc/autograd/FunctionsManual.cpp:2399]: I believe that the input tangent needs to be symmetrised as: `const auto in = 0.5 * (input_tangent + input_tangent.transpose(-2, -1).conj())`.

Denote `Her(n)` the `n x n` Hermitian matrices and `U(n)` the unitary matrices. We have that `eigh : Her(n) -> R^n x U(n)`. This means that the differential of `eigh` goes from the tangent to `Her(n)` to the tangent of `R^n` times the tangent of `U(n)` (at the output matrices). Now, the tangent of `Her(n)` is `Her(n)` itself as  `Her(n)` is just a linear subspace. As such, the input needs to be symmetric for this function to make sense.

Computing `(A + A^H)/2` happens to be the orthogonal projection of a matrix onto the space `Her(n)`. The theorem that formalises all this is the one that says that the differential of a map on an embedded manifold is the differential of the map on the total space restricted to the tangent space of the embedded manifold.

I wrote all this down in a comment in `eigh_backward`.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (Lezcano) [inline torch/csrc/autograd/FunctionsManual.cpp:2419]: `input_tangent` should be projected on to the Hermitian matrices.

**Review by lezcano — COMMENTED** [review]
A few notes on forward mode AD for maps with constrained inputs.

**Comment by IvanYashchuk** [comment]
CI is green now. Thank you, @Lezcano, I improved the PR with your suggestions. Could you take another look please?

@albanD, could you please review this PR?

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:2433]: nit you can return directly here to avoid the extra assignment.

File: torch/testing/_internal/common_methods_invocations.py
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:6856]: Do we have an issue open tracking this? Do we want one if not?

**Review by albanD — APPROVED** [review]
LGTM
I'll wait for @Lezcano final approval and I'll land.

**Review by lezcano — APPROVED** [review]
LGTM modulo @albanD's comments.
The only point is that it might be worth to land this after the support for several outputs in AD mode as, at the moment, this AD is doing twice the same computations. Once for eigvals and another one for eigvectors.

**Comment by albanD** [comment]
Now that this is ready, I think we can just add this to master. And merge the implementation when the multi-output support is added.

File: torch/testing/_internal/common_methods_invocations.py
Comment (IvanYashchuk) [reply torch/testing/_internal/common_methods_invocations.py:6971]: There is no open issue about it. I added a NotImplementedError for complex inputs for now, we'll enable it later.

**Pull request closed without merging**
